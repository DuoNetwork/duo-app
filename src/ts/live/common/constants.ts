export * from '../../../../../duo-admin/src/constants';

export const AC_CTD_STATES = 'custodianStates';
export const AC_CTD_PRICES = 'custodianPrices';
export const AC_ACCOUNT = 'account';
export const AC_BALANCES = 'balances';
export const AC_ADDRESSES = 'addresses';
export const AC_DNM_STATUS = 'dynamoStatus';
export const AC_DMN_HOURLY = 'dynamoHourly';
export const AC_DMN_MINUTELY = 'dynamoMinutely';
export const AC_DMN_PRICES = 'dynamoPrices';
export const AC_REFRESH = 'refresh';
