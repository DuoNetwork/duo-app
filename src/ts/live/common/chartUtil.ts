import moment from 'moment';
import * as CST from '../common/constants';
import { IAcceptedPrice, ICustodianPrice, ICustodianStates, IPriceBar, ISourceData } from './types';

class ChartUtil {
	public interpolate(sourceData: IPriceBar[], isHourly: boolean): IPriceBar[] {
		if (!sourceData.length) return [];
		sourceData.sort((a, b) => a.timestamp - b.timestamp);
		const newSourceData: IPriceBar[] = [];
		const timeStep = isHourly ? 3.6e6 : 60000;
		const source = sourceData[0].source;
		for (let i = 0; i < sourceData.length - 1; i++) {
			newSourceData.push(sourceData[i]);
			const timeIndex = sourceData[i].timestamp;
			const nextTimeIndex = sourceData[i + 1].timestamp;
			const close = sourceData[i].close;
			let newTimeStamp = timeIndex + timeStep;
			while (newTimeStamp < nextTimeIndex) {
				newSourceData.push({
					source: source,
					date: moment(newTimeStamp).format('YYYY-MM-DD'),
					hour: moment(newTimeStamp).format('HH'),
					minute: isHourly ? 0 : Number(moment(newTimeStamp).format('mm')),
					open: close,
					high: close,
					low: close,
					close: close,
					volume: 0,
					timestamp: newTimeStamp
				});
				newTimeStamp += timeStep;
			}
		}
		newSourceData.push(sourceData[sourceData.length - 1]);

		return newSourceData;
	}

	public reset(
		price: IAcceptedPrice[],
		limitUp: number,
		limitDown: number,
		limitPeriod: number
	): IAcceptedPrice[] {
		return price
			.filter(p => p.navB >= limitUp || p.navB <= limitDown || p.navA >= limitPeriod)
			.map(p => ({
				price: p.price,
				navA: 1,
				navB: 1,
				timestamp: p.timestamp + 717
			}));
	}

	public mergeReset(prices: IAcceptedPrice[], resets: IAcceptedPrice[]): IAcceptedPrice[] {
		const all = [...prices, ...resets];
		all.sort((a, b) => a.timestamp - b.timestamp);
		return all;
	}

	public mergeLastToPriceBar(
		pricebars: ISourceData<IPriceBar[]>,
		last: ISourceData<ICustodianPrice>,
		isHourly: boolean
	) {
		CST.EXCHANGES.forEach(src => {
			const pricebar: IPriceBar[] = pricebars[src.toLowerCase()];
			if (!pricebar || !pricebar.length) return;
			const sourceLast: ICustodianPrice = last[src.toLowerCase()];
			const lastBar = pricebar[pricebar.length - 1];
			const dateObj = moment.utc(sourceLast.timestamp);
			if (isHourly && lastBar.date + ' ' + lastBar.hour !== dateObj.format('YYYY-MM-DD HH'))
				pricebar.push({
					source: lastBar.source,
					date: dateObj.format('YYYY-MM-DD'),
					hour: dateObj.format('HH'),
					minute: 0,
					open: lastBar.close,
					high: Math.max(lastBar.close, sourceLast.price),
					low: Math.min(lastBar.close, sourceLast.price),
					close: sourceLast.price,
					volume: 0,
					timestamp: sourceLast.timestamp
				});
			else if (
				!isHourly &&
				lastBar.date + ' ' + lastBar.hour + ' ' + lastBar.minute !==
					dateObj.format('YYYY-MM-DD HH m')
			)
				pricebar.push({
					source: lastBar.source,
					date: dateObj.format('YYYY-MM-DD'),
					hour: dateObj.format('HH'),
					minute: Number(dateObj.format('mm')),
					open: lastBar.close,
					high: Math.max(lastBar.close, sourceLast.price),
					low: Math.min(lastBar.close, sourceLast.price),
					close: sourceLast.price,
					volume: 0,
					timestamp: sourceLast.timestamp
				});
			else {
				lastBar.high = Math.max(lastBar.high, sourceLast.price);
				lastBar.low = Math.min(lastBar.low, sourceLast.price);
				lastBar.close = sourceLast.price;
				lastBar.timestamp = sourceLast.timestamp;
			}
		});
	}

	public mergeLastToPrice(
		prices: IAcceptedPrice[],
		states: ICustodianStates,
		last: ICustodianPrice
	) {
		if (!prices.length) return;
		const lastPrice = prices[prices.length - 1];
		if (lastPrice.timestamp < last.timestamp)
			prices.push({
				price: last.price,
				navA: states.navA,
				navB: states.navB,
				timestamp: last.timestamp
			});
	}
}

const chartUtil = new ChartUtil();
export default chartUtil;
