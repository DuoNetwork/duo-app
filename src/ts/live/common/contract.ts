import infura from '../../../../../duo-admin/src/keys/infura.json';
import ContractUtil from '../../../../../duo-contract-util/src/contractUtil';
import * as CST from './constants';

const provider =
	(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) + '/' + infura.token;
const contract = new ContractUtil(window, '', provider, !__KOVAN__);
export default contract;
