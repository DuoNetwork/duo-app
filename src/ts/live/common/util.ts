import * as d3 from 'd3';
import moment, { DurationInputArg2 } from 'moment';
import * as CST from './constants';
import { ICustodianPrice, IPriceStatus, ISourceData, IStatus } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = this.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'Just Now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' Minutes Ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' Hours Ago';
		else if (diff < 2592000000) return Math.floor(diff / 86400000) + ' Days Ago';
		else return 'Long Time Ago';
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public calculateNav(
		price: number,
		time: number,
		resetPrice: number,
		resetTime: number,
		alpha: number,
		beta: number,
		period: number,
		coupon: number
	) {
		const navParent = (price / resetPrice / beta) * (1 + alpha);

		const navA = 1 + Math.floor((time - resetTime) / 1000 / period) * coupon;
		const navAAdj = navA * alpha;
		if (navParent <= navAAdj) return [navParent / alpha, 0];
		else return [navA, navParent - navAAdj];
	}

	public getDates(length: number, step: number, stepSize: DurationInputArg2, format: string) {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}

	public round(num: number) {
		return +(Math.floor((num + 'e+8') as any) + 'e-8');
	}

	public range(start: number, end: number) {
		const subArray = Array.apply(null, { length: end }).map(Number.call, Number);
		return subArray.slice(start, end);
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public getCurrentPath(location: any) {
		return location.pathname === '/' ? location.pathname + 'beethoven' : location.pathname;
	}

	public getLastPriceFromStatus(status: IStatus[]): ISourceData<ICustodianPrice> {
		// const bitfinex: ICustodianPrice = {
		// 	address: CST.DUMMY_ADDR,
		// 	price: 0,
		// 	timestamp: 0
		// };
		const kraken: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		const gemini: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		const gdax: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		status.forEach(s => {
			// if (s.process === 'PRICE_AWS_PUBLIC_BITFINEX') {
			// 	bitfinex.price = (s as IPriceStatus).price;
			// 	bitfinex.timestamp = (s as IPriceStatus).timestamp;
			// } else
			if (s.process === 'TRADE_AWS_PUBLIC_GEMINI') {
				gemini.price = (s as IPriceStatus).price;
				gemini.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_KRAKEN') {
				kraken.price = (s as IPriceStatus).price;
				kraken.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_GDAX') {
				gdax.price = (s as IPriceStatus).price;
				gdax.timestamp = (s as IPriceStatus).timestamp;
			}
		});

		return {
			// bitfinex,
			kraken,
			gemini,
			gdax
		};
	}
}

const util = new Util();
export default util;
