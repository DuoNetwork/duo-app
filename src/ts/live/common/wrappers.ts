import BeethovenWapper from '../../../../../duo-contract-wrapper/src/BeethovenWapper';
import Web3Wrapper from '../../../../../duo-contract-wrapper/src/Web3Wrapper';
import infura from '../../keys/infura.json';
import * as CST from './constants';

const provider =
	(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) + '/' + infura.token;
export const web3Wrapper = new Web3Wrapper(window, '', provider, !__KOVAN__);
export const beethovenWapper = new BeethovenWapper(web3Wrapper);
