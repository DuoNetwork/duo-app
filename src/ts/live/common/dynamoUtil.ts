import AWS from 'aws-sdk';
import { ScanInput, ScanOutput } from 'aws-sdk/clients/dynamodb';
import devConfig from '../../keys/aws.ui.dev.json';
import liveConfig from '../../keys/aws.ui.live.json';
import { IPriceStatus, IStatus } from '../common/types';
import * as CST from './constants';

export class DynamoUtil {
	private ddb: AWS.DynamoDB;
	constructor() {
		// this.role = role;
		AWS.config.update(__DEV__ ? devConfig : liveConfig);
		this.ddb = new AWS.DynamoDB({ apiVersion: CST.AWS_DYNAMO_API_VERSION });
	}

	public scanData(params: ScanInput): Promise<ScanOutput> {
		return new Promise((resolve, reject) => {
			this.ddb.scan(params, (err, data) => {
				if (err) reject(err);
				else resolve(data);
			});
		});
	}

	public async scanStatus() {
		return this.convertStatus(
			await this.scanData({
				TableName: __DEV__ ? CST.DB_AWS_STATUS_DEV : CST.DB_AWS_STATUS_LIVE
			})
		);
	}

	public convertStatus(status: ScanOutput): IStatus[] {
		if (!status.Items || !status.Items.length) return [];

		const output =  status.Items.map(d => {
			const process = d[CST.DB_ST_PROCESS].S || '';
			const timestamp = Number(d[CST.DB_ST_TS].N);
			if (process.startsWith('PRICE'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_TX_PRICE].N),
					volume: Number(d[CST.DB_TX_AMOUNT].N)
				} as IPriceStatus;
			else if (process.startsWith('FEED'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_HISTORY_PRICE].N),
					volume: Number(d[CST.DB_HISTORY_VOLUME].N)
				} as IPriceStatus;
			else
				return {
					process: process,
					timestamp: timestamp
				};
		});
		output.sort((a, b) => b.timestamp - a.timestamp);
		output.sort((a, b) => a.process.localeCompare(b.process));

		return output;
	}
}

const dynamoUtil = new DynamoUtil();
export default dynamoUtil;
