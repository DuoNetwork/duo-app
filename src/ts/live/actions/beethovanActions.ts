import moment from 'moment';
import dynamoUtil from '../../../../../duo-admin/src/utils/dynamoUtil';
import chartUtil from '../common/chartUtil';
import * as CST from '../common/constants';
import contract from '../common/contract';
import {
	IAcceptedPrice,
	IAccountBalances,
	IAddress,
	IBeethovanAddresses,
	IBeethovanBalances,
	IBeethovanPrices,
	IBeethovanStates,
	IConversion,
	IPrice,
	VoidThunkAction
} from '../common/types';
import util from '../common/util';
import * as web3Actions from './web3Actions';

export function statesUpdate(states: IBeethovanStates) {
	return {
		type: CST.AC_BTV_STATES,
		value: states
	};
}

export function getStates(): VoidThunkAction {
	return async dispatch => {
		const states = await contract.getCustodianStates();
		dispatch(statesUpdate(states));
	};
}

export function pricesUpdate(prices: IBeethovanPrices) {
	return {
		type: CST.AC_BTV_PRICES,
		value: prices
	};
}

export function getPrices(): VoidThunkAction {
	return async dispatch => dispatch(pricesUpdate(await contract.getCustodianPrices()));
}

export function balancesUpdate(balance: IBeethovanBalances) {
	return {
		type: CST.AC_BTV_BALANCES,
		value: balance
	};
}

export function getBalances(): VoidThunkAction {
	return async (dispatch, getState) =>
		dispatch(balancesUpdate(await contract.getBalances(getState().web3.account)));
}

export function allBalancesUpdate(balance: IAccountBalances, index: number) {
	return {
		type: CST.AC_ALL_BALANCES,
		value: {
			[index]: balance
		}
	};
}

export function getAllBalances(start: number, end: number): VoidThunkAction {
	return async dispatch => {
		for (let i = start; i < end; i++)
			contract.getUserAddress(i).then((account: any) => {
				if (account)
					contract.getBalances(account).then(balance =>
						dispatch(
							allBalancesUpdate(
								{
									account: account,
									...balance
								},
								i
							)
						)
					);
			});
	};
}

export function addressesUpdate(addr: IBeethovanAddresses) {
	return {
		type: CST.AC_BTV_ADDRESSES,
		value: addr
	};
}

export function addressPoolUpdate(address: IAddress[]) {
	return {
		type: CST.AC_ADDR_POOL,
		value: address
	};
}

export function getAddresses(): VoidThunkAction {
	return async (dispatch, getState) => {
		dispatch(addressesUpdate(await contract.getCustodianAddresses()));
		const poolLength = getState().beethovan.beethovanStates.addrPoolLength;
		const addrPool: IAddress[] = [];
		for (let i = 0; i < poolLength; i++) {
			const address = await contract.getPoolAddress(i);
			if (address)
				addrPool.push({
					address: address,
					balance: await contract.getEthBalance(address)
				});
		}
		dispatch(addressPoolUpdate(addrPool));
	};
}

export function exchangePricesUpdate(prices: IPrice[]) {
	return {
		type: CST.AC_BTV_EX_PX,
		value: prices
	};
}

export function fetchExchangePrices(): VoidThunkAction {
	return async (dispatch, state) => {
		const source = state().ui.source;
		const period = state().ui.period;
		const start =
			period === 60
				? util.getUTCNowTimestamp() - period * 96 * 60000
				: util.getUTCNowTimestamp() - 400 * 60000;
		dispatch(
			exchangePricesUpdate(
				await dynamoUtil.getPrices(source, period === 60 ? 60 : 1, start, 0, 'ETH|USD')
			)
		);
	};
}

export function acceptedPricesUpdate(acceptedPrices: IAcceptedPrice[]) {
	return {
		type: CST.AC_BTV_ACCEPTED_PX,
		value: acceptedPrices
	};
}

export function fetchAcceptedPrices(contractAddress: string): VoidThunkAction {
	return async (dispatch, getState) => {
		const dates = util.getDates(4, 1, 'day', 'YYYY-MM-DD');
		const priceData = await dynamoUtil.queryAcceptPriceEvent(contractAddress, dates);
		const states = getState().beethovan.beethovanStates;
		dispatch(
			acceptedPricesUpdate(
				chartUtil.mergeReset(
					priceData,
					chartUtil.reset(
						priceData,
						states.limitUpper,
						states.limitLower,
						states.limitPeriodic
					)
				)
			)
		);
	};
}

export function conversionsUpdate(conversions: IConversion[]) {
	return {
		type: CST.AC_BTV_CONVERSIONS,
		value: conversions
	};
}

export function fetchConversions(contractAddress: string): VoidThunkAction {
	return async (dispatch, getState) => {
		const nowTimestamp = util.getUTCNowTimestamp();
		const dates = util.getDates(7, 1, 'day', 'YYYY-MM-DD');
		const account = getState().web3.account;
		const conversion = await dynamoUtil.queryConversionEvent(contractAddress, account, dates);
		(await dynamoUtil.queryUIConversionEvent(contractAddress, account)).forEach(uc => {
			if (
				!dates.includes(moment.utc(uc.timestamp).format('YYYY-MM-DD')) ||
				conversion.some(c => c.transactionHash === uc.transactionHash) ||
				(uc.pending && nowTimestamp - uc.timestamp > CST.PENDING_TX_TIMEOUT)
			)
				dynamoUtil.deleteUIConversionEvent(account, uc);
			else conversion.push(uc);
		});
		conversion.sort((a, b) => -a.timestamp + b.timestamp);
		dispatch(conversionsUpdate(conversion));
	};
}

export function refresh(): VoidThunkAction {
	return async dispatch => {
		await dispatch(getStates());
		dispatch(getPrices());
		dispatch(getBalances());
		dispatch(getAddresses());
		dispatch(fetchExchangePrices());
		dispatch(fetchAcceptedPrices(contract.custodianAddr));
		dispatch(fetchConversions(contract.custodianAddr));
		window.setInterval(async () => {
			await dispatch(getStates());
			dispatch(getPrices());
			dispatch(getBalances());
			dispatch(getAddresses());
			dispatch(fetchExchangePrices());
			dispatch(fetchAcceptedPrices(contract.custodianAddr));
			dispatch(fetchConversions(contract.custodianAddr));
		}, 60000);
	};
}

export function adminActions(): VoidThunkAction {
	return async dispatch => {
		await dispatch(getStates());
		dispatch(getAddresses());
		dispatch(web3Actions.getAccount());
		window.setInterval(async () => {
			await dispatch(getStates());
			dispatch(getAddresses());
			dispatch(web3Actions.getAccount());
		}, 60000);
	};
}

export function userActions(start: number, end: number): VoidThunkAction {
	return async dispatch => {
		await dispatch(getStates());
		dispatch(getAllBalances(start, end));
		window.setInterval(async () => {
			await dispatch(getStates());
			dispatch(getAllBalances(start, end));
		}, 60000);
	};
}

export function statusActions(): VoidThunkAction {
	return async dispatch => {
		await dispatch(getStates());
		window.setInterval(async () => {
			await dispatch(getStates());
		}, 60000);
	};
}
