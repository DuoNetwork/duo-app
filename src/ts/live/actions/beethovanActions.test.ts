import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import dynamoUtil from '../../../../../duo-admin/src/utils/dynamoUtil';
import * as CST from '../common/constants';
import util from '../common/util';
import { beethovanWapper } from '../common/wrappers';
import * as beethovanActions from './beethovanActions';

const mockStore = configureMockStore([thunk]);

describe('actions', () => {
	test('statesUpdate', () => {
		expect(beethovanActions.statesUpdate({ test: 'test' } as any)).toMatchSnapshot();
	});

	test('getStates', () => {
		const store: any = mockStore({});
		beethovanWapper.getStates = jest.fn(() =>
			Promise.resolve({
				test: 'test'
			})
		);
		store.dispatch(beethovanActions.getStates());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('addressesUpdate', () => {
		expect(beethovanActions.addressesUpdate({ test: 'test' } as any)).toMatchSnapshot();
	});

	test('getAddresses', () => {
		const store: any = mockStore({ beethovan: { beethovanStates: { addrPoolLength: 1 } } });
		beethovanWapper.getAddresses = jest.fn(() => Promise.resolve({ test: 'test' }));
		store.dispatch(beethovanActions.getAddresses());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('exchangePricesUpdate', () => {
		expect(beethovanActions.exchangePricesUpdate(['test'] as any)).toMatchSnapshot();
	});

	test('fetchExchangePrices 60', () => {
		util.getUTCNowTimestamp = jest.fn(() => 1234567890);
		const store: any = mockStore({
			ui: {
				period: 60,
				source: 'test'
			}
		});
		dynamoUtil.getPrices = jest.fn(
			(src: string, period: number, start: number, end: number, pair: string) =>
				Promise.resolve([src, period, start, end, pair])
		);
		store.dispatch(beethovanActions.fetchExchangePrices());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('fetchExchangePrices 5', () => {
		util.getUTCNowTimestamp = jest.fn(() => 1234567890);
		const store: any = mockStore({
			ui: {
				period: 5,
				source: 'test'
			}
		});
		dynamoUtil.getPrices = jest.fn(
			(src: string, period: number, start: number, end: number, pair: string) =>
				Promise.resolve([src, period, start, end, pair])
		);
		store.dispatch(beethovanActions.fetchExchangePrices());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('acceptedPricesUpdate', () => {
		expect(beethovanActions.acceptedPricesUpdate(['test'] as any)).toMatchSnapshot();
	});

	test('fetchAcceptedPrices', () => {
		const store: any = mockStore({
			beethovan: {
				states: { limitUpper: 2, limitLower: 0.25, limitPeriodic: 1.035 }
			}
		});
		dynamoUtil.queryAcceptPriceEvent = jest.fn(() => Promise.resolve(['test']));
		store.dispatch(beethovanActions.fetchAcceptedPrices(CST.DUMMY_ADDR));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('conversionsUpdate', () => {
		expect(beethovanActions.conversionsUpdate(['test'] as any)).toMatchSnapshot();
	});

	test('fetchConversions', () => {
		const store: any = mockStore({ web3: { account: CST.DUMMY_ADDR } });
		util.getDates = jest.fn(() => ['1970-01-15']);
		dynamoUtil.queryConversionEvent = jest.fn(() =>
			Promise.resolve([
				{
					transactionHash: 'aaa'
				}
			])
		);
		dynamoUtil.queryUIConversionEvent = jest.fn(() =>
			Promise.resolve([
				{
					transactionHash: 'aaa',
					timestamp: 1234567890
				},
				{
					transactionHash: 'bbb',
					timestamp: 1234567890
				},
				{
					transactionHash: 'ccc',
					timestamp: 0
				}
			])
		);
		dynamoUtil.deleteUIConversionEvent = jest.fn(() => Promise.resolve());
		store.dispatch(beethovanActions.fetchConversions(CST.DUMMY_ADDR));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				expect(
					(dynamoUtil.deleteUIConversionEvent as jest.Mock<Promise<void>>).mock.calls
						.length
				).toBe(2);
				expect(
					(dynamoUtil.deleteUIConversionEvent as jest.Mock<Promise<void>>).mock
						.calls[0][1]
				).toMatchSnapshot();
				expect(
					(dynamoUtil.deleteUIConversionEvent as jest.Mock<Promise<void>>).mock
						.calls[1][1]
				).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('refresh', () => {
		util.getDates = jest.fn(() => ['1970-01-15']);
		dynamoUtil.queryConversionEvent = jest.fn(() =>
			Promise.resolve([
				{
					transactionHash: 'aaa'
				}
			])
		);
		dynamoUtil.queryUIConversionEvent = jest.fn(() =>
			Promise.resolve([
				{
					transactionHash: 'aaa',
					timestamp: 1234567890
				},
				{
					transactionHash: 'bbb',
					timestamp: 1234567890
				},
				{
					transactionHash: 'ccc',
					timestamp: 0
				}
			])
		);
		dynamoUtil.deleteUIConversionEvent = jest.fn(() => Promise.resolve());
		util.getUTCNowTimestamp = jest.fn(() => 1234567890);
		dynamoUtil.getPrices = jest.fn(
			(src: string, period: number, start: number, end: number, pair: string) =>
				Promise.resolve([src, period, start, end, pair])
		);
		const store: any = mockStore({
			web3: { account: CST.DUMMY_ADDR },
			beethovan: {
				states: {
					addrPoolLength: 1,
					limitUpper: 2,
					limitLower: 0.25,
					limitPeriodic: 1.035
				}
			},
			ui: { period: 5, source: 'test' }
		});
		dynamoUtil.queryAcceptPriceEvent = jest.fn(() => Promise.resolve(['test']));
		beethovanWapper.getStates = jest.fn(() => Promise.resolve({ test: 'test' }));
		store.dispatch(beethovanActions.refresh('custodian'));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('subscriptionUpdate', () => {
		expect(beethovanActions.subscriptionUpdate(123)).toMatchSnapshot();
	});

	test('subscribe', () => {
		window.setInterval = jest.fn(() => 123);
		util.getDates = jest.fn(() => ['1970-01-15']);
		dynamoUtil.queryConversionEvent = jest.fn(() =>
			Promise.resolve([
				{
					transactionHash: 'aaa'
				}
			])
		);
		dynamoUtil.queryUIConversionEvent = jest.fn(() =>
			Promise.resolve([
				{
					transactionHash: 'aaa',
					timestamp: 1234567890
				},
				{
					transactionHash: 'bbb',
					timestamp: 1234567890
				},
				{
					transactionHash: 'ccc',
					timestamp: 0
				}
			])
		);
		dynamoUtil.deleteUIConversionEvent = jest.fn(() => Promise.resolve());
		util.getUTCNowTimestamp = jest.fn(() => 1234567890);
		dynamoUtil.getPrices = jest.fn(
			(src: string, period: number, start: number, end: number, pair: string) =>
				Promise.resolve([src, period, start, end, pair])
		);
		const store: any = mockStore({
			web3: { account: CST.DUMMY_ADDR },
			beethovan: {
				states: {
					addrPoolLength: 1,
					limitUpper: 2,
					limitLower: 0.25,
					limitPeriodic: 1.035
				}
			},
			ui: { period: 5, source: 'test' }
		});
		dynamoUtil.queryAcceptPriceEvent = jest.fn(() => Promise.resolve(['test']));
		beethovanWapper.getStates = jest.fn(() => Promise.resolve({ test: 'test' }));
		store.dispatch(beethovanActions.subscribe('custodian'));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
});
