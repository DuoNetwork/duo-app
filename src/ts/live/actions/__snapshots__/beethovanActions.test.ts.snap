// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions acceptedPricesUpdate 1`] = `
Object {
  "type": "beethovanAcceptedPrices",
  "value": Array [
    "test",
  ],
}
`;

exports[`actions addressPoolUpdate 1`] = `
Object {
  "type": "addressPool",
  "value": Array [
    Object {
      "test": "test",
    },
  ],
}
`;

exports[`actions addressesUpdate 1`] = `
Object {
  "type": "beethovanAddresses",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions allBalancesUpdate 1`] = `
Object {
  "type": "allBalances",
  "value": Object {
    "123": Object {
      "test": "test",
    },
  },
}
`;

exports[`actions balancesUpdate 1`] = `
Object {
  "type": "beethovanBalances",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions conversionsUpdate 1`] = `
Object {
  "type": "beethovanConversions",
  "value": Array [
    "test",
  ],
}
`;

exports[`actions exchangePricesUpdate 1`] = `
Object {
  "type": "beethovanExchangePrices",
  "value": Array [
    "test",
  ],
}
`;

exports[`actions fetchAcceptedPrices 1`] = `
Array [
  Object {
    "type": "beethovanAcceptedPrices",
    "value": Array [
      "test",
    ],
  },
]
`;

exports[`actions fetchConversions 1`] = `
Array [
  Object {
    "type": "beethovanConversions",
    "value": Array [
      Object {
        "transactionHash": "aaa",
      },
      Object {
        "timestamp": 1234567890,
        "transactionHash": "bbb",
      },
    ],
  },
]
`;

exports[`actions fetchConversions 2`] = `
Object {
  "timestamp": 1234567890,
  "transactionHash": "aaa",
}
`;

exports[`actions fetchConversions 3`] = `
Object {
  "timestamp": 0,
  "transactionHash": "ccc",
}
`;

exports[`actions fetchExchangePrices 5 1`] = `
Array [
  Object {
    "type": "beethovanExchangePrices",
    "value": Array [
      "test",
      1,
      1210567890,
      0,
      "ETH|USD",
    ],
  },
]
`;

exports[`actions fetchExchangePrices 60 1`] = `
Array [
  Object {
    "type": "beethovanExchangePrices",
    "value": Array [
      "test",
      60,
      888967890,
      0,
      "ETH|USD",
    ],
  },
]
`;

exports[`actions getAddresses 1`] = `
Array [
  Object {
    "type": "beethovanAddresses",
    "value": Object {
      "test": "test",
    },
  },
  Object {
    "type": "addressPool",
    "value": Array [
      Object {
        "address": "0x0",
        "balance": 123,
      },
    ],
  },
]
`;

exports[`actions getAllBalances 1`] = `
Array [
  Object {
    "type": "allBalances",
    "value": Object {
      "123": Object {
        "account": "0x0",
        "test": "test",
      },
    },
  },
  Object {
    "type": "allBalances",
    "value": Object {
      "124": Object {
        "account": "0x0",
        "test": "test",
      },
    },
  },
]
`;

exports[`actions getBalances 1`] = `
Array [
  Object {
    "type": "beethovanBalances",
    "value": Object {
      "test": "test",
    },
  },
]
`;

exports[`actions getPrices 1`] = `
Array [
  Object {
    "type": "beethovanPrices",
    "value": Object {
      "last": "last",
      "reset": "reset",
    },
  },
]
`;

exports[`actions getStates 1`] = `
Array [
  Object {
    "type": "beethovanStates",
    "value": Object {
      "test": "test",
    },
  },
]
`;

exports[`actions pricesUpdate 1`] = `
Object {
  "type": "beethovanPrices",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions statesUpdate 1`] = `
Object {
  "type": "beethovanStates",
  "value": Object {
    "test": "test",
  },
}
`;
